src/Bicep.Core/Analyzers/Linter/Rules/NoUnusedRuleBase.cs

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Immutable;
+using Bicep.Core.CodeAction;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Parsing;
+using Bicep.Core.Syntax;
+using Bicep.Core.Text;
+
+namespace Bicep.Core.Analyzers.Linter.Rules;
+
+public abstract class NoUnusedRuleBase : LinterRuleBase
+{
+    protected const string MissingName = "<missing>";
+    private readonly string type;
+
+    protected NoUnusedRuleBase(string type, string code, string description, Uri? docUri = null, DiagnosticLevel diagnosticLevel = DiagnosticLevel.Warning, DiagnosticLabel? diagnosticLabel = null) :
+        base(code, description, docUri, diagnosticLevel, diagnosticLabel)
+    {
+        this.type = type;
+    }
+
+    protected AnalyzerFixableDiagnostic CreateRemoveUnusedDiagnosticForSpan(string name, IdentifierSyntax nameSyntax, SyntaxBase declaringSyntax, ImmutableArray<int> lineStarts)
+    {
+        var span = GetSpanForRow(declaringSyntax, nameSyntax, lineStarts);
+        var codeFix = new CodeFix($"Remove unused {type}", true, CodeFixKind.QuickFix, new CodeReplacement(span, String.Empty));
+        var fixableDiagnosticForSpan = CreateFixableDiagnosticForSpan(nameSyntax.Span, codeFix, name);
+        return fixableDiagnosticForSpan;
+    }
+
+    private static TextSpan GetSpanForRow(SyntaxBase declaringSyntax, IdentifierSyntax identifierSyntax, ImmutableArray<int> lineStarts)
+    {
+        var spanPosition = declaringSyntax.Span.Position;
```

Comment:
Here's my suggestion for making this function more robust with newlines, leading spaces and comments, and multi line comments:
```csharp
private static TextSpan GetSpanForRow(ProgramSyntax programSyntax, SyntaxBase declaringSyntax, IdentifierSyntax identifierSyntax, ImmutableArray<int> lineStarts)
{
    // Find the first & last token in the statement
    var startToken = declaringSyntax.TryFindMostSpecificNodeInclusive(declaringSyntax.Span.Position, x => x is Token) as Token;
    var endToken = declaringSyntax.TryFindMostSpecificNodeInclusive(declaringSyntax.Span.Position + declaringSyntax.Span.Length, x => x is Token) as Token;

    // This shouldn't happen, but if it does - fall back to just replacing the statement only
    if (startToken is null || endToken is null)
    {
        return declaringSyntax.Span;
    }

    // If we have leading or trailing trivia (whitespace or comments), take the outermost trivia
    var startPosSpan = startToken.LeadingTrivia.FirstOrDefault()?.Span ?? startToken.Span;
    var endPosSpan = endToken.TrailingTrivia.LastOrDefault()?.Span ?? endToken.Span;

    // If we have a trailing newline, include it in the calculation so that it is removed
    var followingToken = programSyntax.TryFindMostSpecificNodeInclusive(endPosSpan.Position + endPosSpan.Length, x => x is Token);
    if (followingToken is Token { Type: TokenType.NewLine } newLineToken)
    {
        endPosSpan = newLineToken.Span;
    }

    return TextSpan.Between(startPosSpan, endPosSpan);
}
```

Ultimately, I think it's necessary to use the syntax tree to understand the spaces/comments/newlines, or it'll be tricky to handle the edge cases properly.

Reply:
Thanks. I implemented your suggestion.